Updated to clarify what a checkpoint is and how to create it (#790)
EventStore/documentation: docs/clients/grpc/subscriptions.md

You provide an event handler and an optional starting point to the subscription. The handler is called for each event from the starting point onward.

> If events already exist, the handler will be called for each event one by one until it reaches the end of the stream. The server will then notify the handler whenever a new event appears.

Reviewer: ```suggestion
If events exist, the handler processes each event sequentially until it reaches the end of the stream. The server notifies the handler when a new event appears.
```

### Subscribing to `$all`

Subscribing to `$all` is similar to subscribing to a single stream. The handler will be called for every event appended after the starting position.

@[code{subscribe-to-all}](@grpc:subscribing_to_stream.py;subscribing-to-streams.js;subscribing-to-streams.ts;subscribing_to_stream/SubscribingToStream.java;subscribing-to-streams/Program.cs;subscribingToStream.go;subscribing_to_stream.rs)

## Subscribing from a specific position

The previous examples subscribed to the stream from the beginning. This called the handler for every event in the stream and then waited for new events.

> Both stream and $all subscriptions accept a starting position if you want to read from a specific point onward. If events already exist at the position you subscribe to, they will be read on the server side and sent to the subscription.

Reviewer: ```suggestion
Both stream and `$all` subscriptions let you specify a starting position for reading. If events exist at that position, the server reads and sends them to the subscription.
```

## Server-side filtering

EventStoreDB allows you to filter the events whilst subscribing to the `$all` stream to only receive the events you care about.

You can filter by event type or stream name using a regular expression or a prefix. Server-side filtering is currently only available on the `$all` stream.

::: tip

> Server-side filtering was introduced as a simpler alternative to projections. You should consider filtering before creating a projection to include the events you care about.

Reviewer: ```suggestion
Server-side filtering was introduced as a simpler alternative to projections. Consider filtering before creating a projection to include the events you care about.
```

This will subscribe to any event that begins with `user` or `company`.

### Filtering by stream name

If you want to subscribe to a stream only with a given name, you have two options: you can use a regular expression or a prefix.

#### Filtering by prefix

If you want to filter by prefix, pass in a `SubscriptionFilterOptions` to the subscription with an `StreamFilter.Prefix`.

@[code{stream-prefix}](@grpc:server_side_filtering.py;server-side-filtering.js;server-side-filtering.ts;server_side_filtering/ServerSideFiltering.java;server-side-filtering/Program.cs;serverSideFiltering.go;server_side_filtering.rs)

This will only subscribe to all streams with a name that begins with `user-`.

#### Filtering by regular expression

If you want to subscribe to multiple streams, providing a regular expression might be better.

@[code{stream-regex}](@grpc:server_side_filtering.py;server-side-filtering.js;server-side-filtering.ts;server_side_filtering/ServerSideFiltering.java;server-side-filtering/Program.cs;serverSideFiltering.go;server_side_filtering.rs)

This will subscribe to any stream with a name that begins with `account` or `savings`.

## Checkpointing

> When a catch-up subscription is used to process an `$all` stream containing many events, the last thing you want is for your application to crash midway, forcing you to restart from the beginning.

Reviewer: ```suggestion
When processing an `$all` stream with a catch-up subscription, ensure your application stays stable to avoid restarting from the beginning.
```
